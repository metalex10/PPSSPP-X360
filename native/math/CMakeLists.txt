set(SRCS
  lin/matrix4x4.cpp
  lin/vec3.cpp
  lin/quat.cpp
  lin/aabb.cpp
	curves.cpp
	math_util.cpp
  )

set(SRCS ${SRCS})

add_library(lin STATIC ${SRCS})

if(UNIX)
  add_definitions(-fPIC)
endif(UNIX)
