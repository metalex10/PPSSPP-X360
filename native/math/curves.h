#pragma once

#include <base/basictypes.h>

// Easy curve computation for fades etc.


// output range: [0.0, 1.0]
float linearInOut(int t, int fadeInLength, int solidLength, int fadeOutLength);
float linearIn(int t, int fadeInLength);
float linearOut(int t, int fadeInLength);

// smooth operator [0, 1] -> [0, 1]
float ease(float val);
float ease(int t, int fadeLength);

// need a bouncy ease

// waveforms [0, 1]
float sawtooth(int t, int period);

// output range: -1.0 to 1.0
float passWithPause(int t, int fadeInLength, int pauseLength, int fadeOutLength);
