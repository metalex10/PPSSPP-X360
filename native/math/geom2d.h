#pragma once

#include <cmath>

struct Point {
	Point() {}
	Point(float x_, float y_) : x(x_), y(y_) {}

	float x;
	float y;

	float distanceTo(const Point &other) const {
		float dx = other.x - x, dy = other.y - y;
		return sqrtf(dx*dx + dy*dy);
	}

	/*
	FocusDirection directionTo(const Point &other) const {
		int angle = atan2f(other.y - y, other.x - x) / (2 * M_PI) - 0.125;

	}*/
};


// Resolved bounds on screen after layout.
struct Bounds {
	Bounds() : x(0), y(0), w(0), h(0) {}
	Bounds(float x_, float y_, float w_, float h_) : x(x_), y(y_), w(w_), h(h_) {}

	bool Contains(float px, float py) const {
		return (px >= x && py >= y && px < x + w && py < y + h);
	}

	bool Intersects(const Bounds &other) const {
		return !(x > other.x2() || x2() < other.x || y > other.y2() || y2() < other.y);
	}

	void Clip(const Bounds &clipTo) {
		if (x < clipTo.x) {
			w -= clipTo.x - x;
			x = clipTo.x;
		}
		if (y < clipTo.y) {
			h -= clipTo.y - y;
			y = clipTo.y;
		}
		if (x2() > clipTo.x2()) {
			w = clipTo.x2() - x;
		}
		if (y2() > clipTo.y2()) {
			h = clipTo.y2() - y;
		}
	}

	float x2() const { return x + w; }
	float y2() const { return y + h; }
	float centerX() const { return x + w * 0.5f; }
	float centerY() const { return y + h * 0.5f; }
	Point Center() const {
		return Point(centerX(), centerY());
	}

	float x;
	float y;
	float w;
	float h;
};


