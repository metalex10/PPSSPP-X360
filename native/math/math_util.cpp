#include "math/math_util.h"
#include <stdlib.h>

#if defined(__ARM_ARCH_7A__) && !defined(BLACKBERRY)

void EnableFZ()
{
	int x;
	asm(
		"fmrx %[result],FPSCR \r\n"
		"orr %[result],%[result],#16777216 \r\n"
		"fmxr FPSCR,%[result]"
		:[result] "=r" (x) : :
	);
	//printf("ARM FPSCR: %08x\n",x);
}
#else
void EnableFZ()
{
	// TODO
}
#endif
