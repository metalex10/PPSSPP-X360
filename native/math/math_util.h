#ifndef _MATHUTILS_H
#define _MATHUTILS_H

#include <cmath>
#include <cstring>

#include "base/basictypes.h"

inline float sqr(float f)	{return f*f;}
inline float sqr_signed(float f) {return f<0 ? -f*f : f*f;}

typedef unsigned short float16;

// This ain't a 1.5.10 float16, it's a stupid hack format where we chop 16 bits off a float.
// This choice is subject to change. Don't think I'm using this for anything at all now anyway.
// DEPRECATED
inline float16 FloatToFloat16(float x) {
	int ix;
	memcpy(&ix, &x, sizeof(float));
	return ix >> 16;
}

inline float Float16ToFloat(float16 ix) {
	float x;
	memcpy(&x, &ix, sizeof(float));
	return x;
}

inline uint32_t RoundUpToPowerOf2(uint32_t v) {
	v--;
	v |= v >> 1;
	v |= v >> 2;
	v |= v >> 4;
	v |= v >> 8;
	v |= v >> 16;
	v++;
	return v;
}

#define PI 3.141592653589793f
#ifndef M_PI
#define M_PI 3.141592653589793f
#endif

// The stuff in this file is from all over the web, esp. dspmusic.org. I think it's all public domain.
// In any case, very little of it is used anywhere at the moment.

// Calculate pseudo-random 32 bit number based on linear congruential method. 
void SetSeed(unsigned int seed);
unsigned int GenerateRandomNumber();
inline float GenerateRandomFloat01() {
	return (float)((double)GenerateRandomNumber() / 0xFFFFFFFF);
}
inline float GenerateRandomSignedFloat() {
	return (float)((double)GenerateRandomNumber() / 0x80000000) - 1.0f;
}


inline float GaussRand()
{
	float R1 = GenerateRandomFloat01();
	float R2 = GenerateRandomFloat01();

	float X = sqrtf(-2.0f * logf(R1)) * cosf(2.0f * PI * R2);
	if (X > 4.0f) X = 4.0f;
	if (X < -4.0f) X = -4.0f;
	return X;
}

// Accuracy unknown
inline double atan_fast(double x) {
	return (x / (1.0 + 0.28 * (x * x)));
}


// linear -> dB conversion
inline float lin2dB(float lin) {
	const float LOG_2_DB = 8.6858896380650365530225783783321f;	// 20 / ln( 10 )
	return logf(lin) * LOG_2_DB;
}

// dB -> linear conversion
inline float dB2lin(float dB) {
	const float DB_2_LOG = 0.11512925464970228420089957273422f;	// ln( 10 ) / 20
	return expf(dB * DB_2_LOG);
}

inline bool my_isinf(float f) {
	union {
		float f;
		uint32_t u;
	} f2u;
	f2u.f = f;
	return f2u.u == 0x7f800000 ||
		f2u.u == 0xff800000;
}

inline bool my_isnan(float f) {
	union {
		float f;
		uint32_t u;
	} f2u;
	f2u.f = f;
	// NaNs have non-zero mantissa
	return ((f2u.u & 0x7F800000) == 0x7F800000) && (f2u.u & 0x7FFFFF);
}

inline bool my_isnanorinf(float f) {
	union {
		float f;
		uint32_t u;
	} f2u;
	f2u.f = f;
	// NaNs have non-zero mantissa, infs have zero mantissa. That is, we just ignore the mantissa here.
	return ((f2u.u & 0x7F800000) == 0x7F800000);
}

inline int is_even(float d) {
	float int_part;
	modff(d / 2.0f, &int_part);
	return 2.0f * int_part == d;
}

// Rounds *.5 to closest even number
inline double round_ieee_754(double d) {
	float i = floor(d);
	d -= i;
	if (d < 0.5f)
		return i;
	if (d > 0.5f)
		return i + 1.0f;
	if (is_even(i))
		return i;
	return i + 1.0f;
}

union FP32 {
	uint32_t u;
	float f;
};

struct FP16 {
	uint16_t u;
};

// magic code from ryg: http://fgiesen.wordpress.com/2012/03/28/half-to-float-done-quic/
// See also SSE2 version: https://gist.github.com/rygorous/2144712
inline FP32 half_to_float_fast5(FP16 h)
{
	static const FP32 magic = { (127 + (127 - 15)) << 23 };
	static const FP32 was_infnan = { (127 + 16) << 23 };
	FP32 o;
	o.u = (h.u & 0x7fff) << 13;     // exponent/mantissa bits
	o.f *= magic.f;                 // exponent adjust
	if (o.f >= was_infnan.f)        // make sure Inf/NaN survive
		o.u |= 255 << 23;
	o.u |= (h.u & 0x8000) << 16;    // sign bit
	return o;
}

inline float ExpandHalf(uint16_t half) {
	FP16 fp16;
	fp16.u = half;
	FP32 fp = half_to_float_fast5(fp16);
	return fp.f;
}

// More magic code: https://gist.github.com/rygorous/2156668
inline FP16 float_to_half_fast3(FP32 f)
{
	static const FP32 f32infty = { 255 << 23 };
	static const FP32 f16infty = { 31 << 23 };
	static const FP32 magic = { 15 << 23 };
	static const uint32_t sign_mask = 0x80000000u;
	static const uint32_t round_mask = ~0xfffu;
	FP16 o = { 0 };

	uint32_t sign = f.u & sign_mask;
	f.u ^= sign;

	if (f.u >= f32infty.u) // Inf or NaN (all exponent bits set)
		o.u = (f.u > f32infty.u) ? 0x7e00 : 0x7c00; // NaN->qNaN and Inf->Inf
	else // (De)normalized number or zero
	{
		f.u &= round_mask;
		f.f *= magic.f;
		f.u -= round_mask;
		if (f.u > f16infty.u) f.u = f16infty.u; // Clamp to signed infinity if overflowed

		o.u = f.u >> 13; // Take the bits!
	}

	o.u |= sign >> 16;
	return o;
}

inline uint16_t ShrinkToHalf(float full) {
	FP32 fp32;
	fp32.f = full;
	FP16 fp = float_to_half_fast3(fp32);
	return fp.u;
}

// FPU control.
void EnableFZ();

#endif
