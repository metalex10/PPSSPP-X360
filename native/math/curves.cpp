
#include <math.h>

#include "math/math_util.h"
#include "curves.h"

float linearInOut(int t, int fadeInLength, int solidLength, int fadeOutLength) {
	if (t < 0) return 0;
	if (t < fadeInLength) {
		return (float)t / fadeInLength;
	}
	t -= fadeInLength;
	if (t < solidLength) {
		return 1.0f;
	}
	t -= solidLength;
	if (t < fadeOutLength) {
		return 1.0f - (float)t / fadeOutLength;
	}
	return 0.0f;
}

float linearIn(int t, int fadeInLength) {
	if (t < 0) return 0;
	if (t < fadeInLength) {
		return (float)t / fadeInLength;
	}
	return 1.0f;
}

float linearOut(int t, int fadeOutLength) {
	return 1.0f - linearIn(t, fadeOutLength);
}

float ease(float val) {
	if (val > 1.0f) return 1.0f;
	if (val < 0.0f) return 0.0f;
	return ((-cosf(val * PI)) + 1.0f) * 0.5;
}

float ease(int t, int fadeLength)
{
	if (t < 0) return 0.0f;
	if (t >= fadeLength) return 1.0f;
	return ease((float)t / (float)fadeLength);
}

float sawtooth(int t, int period) {
	return (t % period) * (1.0f / (period - 1));
}

float passWithPause(int t, int fadeInLength, int pauseLength, int fadeOutLength)
{
	if (t < fadeInLength) {
		return -1.0f + (float)t / fadeInLength;
	}
	t -= fadeInLength;
	if (t < pauseLength) {
		return 0.0f;
	}
	t -= pauseLength;
	if (t < fadeOutLength) {
		return (float)t / fadeOutLength;
	}
	return 1.0f;
}
