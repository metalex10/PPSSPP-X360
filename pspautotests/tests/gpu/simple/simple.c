#include <common.h>

#include <pspkernel.h>
#include <pspdisplay.h>
#include <pspdebug.h>
#include <stdlib.h>
#include <stdio.h>
#include <math.h>
#include <string.h>
#include <time.h>

#include <pspgu.h>
#include <pspgum.h>
#include <pspdisplay.h>

static unsigned int __attribute__((aligned(16))) list[262144];

typedef struct {
	char u, v;
	unsigned int color;
	char x, y, z;
} VertexType1;

typedef struct {
	char u, v;
	char x, y, z;
} VertexType2;

typedef struct {
	unsigned int color;
	float x, y, z;
} VertexType3;

#define BUILD_COL(COLOR) 0xFF##COLOR

unsigned int clut16[16] = {
	BUILD_COL(000000),
	BUILD_COL(111111),
	BUILD_COL(222222),
	BUILD_COL(333333),
	BUILD_COL(444444),
	BUILD_COL(555555),
	BUILD_COL(666666),
	BUILD_COL(777777),
	BUILD_COL(888888),
	BUILD_COL(999999),
	BUILD_COL(AAAAAA),
	BUILD_COL(BBBBBB),
	BUILD_COL(CCCCCC),
	BUILD_COL(DDDDDD),
	BUILD_COL(EEEEEE),
	BUILD_COL(FFFFFF),
};

unsigned char tex4_4[4 * 4] = {
	 0,  1,  2,  3,
	 4,  5,  6,  7,
	 8,  9, 10, 11,
	12, 13, 14, 15
};

void dumpPixels(int dx, int dy, int w, int h) {
	int x, y;
	unsigned int *pixels;
	int bufferWidth;
	int pixelformat;
	sceDisplayGetFrameBuf((void **)&pixels, &bufferWidth, &pixelformat, PSP_DISPLAY_SETBUF_IMMEDIATE);
	
	for (x = 0; x < w; x++) {
		for (y = 0; y < h; y++) {
			printf("%04X,", pixels[(bufferWidth * (y + dy)) + (x + dx)]);
		}
		printf("\n");
	}
}

VertexType1 buildVertexType1(unsigned int color, char u, char v, char x, char y, char z) {
	VertexType1 vt;
	vt.color = color;
	vt.u = u; vt.v = v;
	vt.x = x; vt.y = y; vt.z = z;
	return vt;
}

VertexType2 buildVertexType2(char u, char v, char x, char y, char z) {
	VertexType2 vt;
	vt.u = u; vt.v = v;
	vt.x = x; vt.y = y; vt.z = z;
	return vt;
}

/*
VertexType3 buildVertexType3(float x, float y, float z) {
	VertexType3 vt;
	vt.x = x; vt.y = y; vt.z = z;
	return vt;
}
*/

// Every vertex has to be aligned to the maxium size of all of its component.
/**
 * Checks that engine is considering the vertex alignment.
 * Try to avoid regressions since the fix of the alignment at revision 252.
 */
void testVertexAlignment() {
	int vertexCount = 2;
	VertexType1* vertices = (VertexType1 *)sceGuGetMemory(vertexCount * sizeof(VertexType1));
	vertices[0] = buildVertexType1(0xFF0000FF, 0, 0, 0, 0, 0);
	vertices[1] = buildVertexType1(0xFF0000FF, 4, 4, 4, 4, 0);

	printf("testVertexAlignment\n");
	printf("Struct Size: %d\n", sizeof(VertexType1)); // 12

	sceGuStart(GU_DIRECT,list);
	sceGuClear(GU_COLOR_BUFFER_BIT | GU_DEPTH_BUFFER_BIT);
	{
		sceGuEnable(GU_TEXTURE_2D);
		sceGuClutMode(GU_PSM_8888, 0, 0xFF, 0); // 32-bit palette
		sceGuClutLoad((16 / 8), clut16); // upload 32*8 entries (256)
		sceGuTexMode(GU_PSM_T8, 0, 0, 0); // 8-bit image
		sceGuTexImage(0, 4, 4, 4, tex4_4);
		sceGuTexFunc(GU_TFX_REPLACE, GU_TCC_RGB);
		sceGuTexFilter(GU_LINEAR, GU_LINEAR);
		sceGuTexScale(1.0f, 1.0f);
		sceGuTexOffset(0.0f, 0.0f);
		//sceGuAmbientColor(0xffffffff);
	
		sceGuDrawArray(GU_SPRITES, GU_TEXTURE_8BIT | GU_COLOR_8888 | GU_VERTEX_8BIT | GU_TRANSFORM_2D, vertexCount, 0, vertices);
	}
	sceGuFinish();
	sceGuSync(0, 0);
	sceGuSwapBuffers();
	
	dumpPixels(0, 0, 4, 4);
}

void testColorAdd() {
	int vertexCount = 2;
	VertexType2* vertices = (VertexType2 *)sceGuGetMemory(vertexCount * sizeof(VertexType2));
	vertices[0] = buildVertexType2(0, 0, 0, 0, 0);
	vertices[1] = buildVertexType2(4, 4, 4, 4, 0);

	printf("testColorAdd\n");
	
	sceGuStart(GU_DIRECT, list);
	sceGuClear(GU_COLOR_BUFFER_BIT | GU_DEPTH_BUFFER_BIT);
	{
		sceGuDisable(GU_TEXTURE_2D);
		sceGuColor(0xff0000ff);
		sceGuDrawArray(GU_SPRITES, GU_TEXTURE_8BIT | GU_VERTEX_8BIT | GU_TRANSFORM_2D, vertexCount, 0, vertices);
	}
	sceGuFinish();
	sceGuSync(0, 0);

	sceGuStart(GU_DIRECT, list);
	{
		//sceGuColor(0x00000000);

		sceGuEnable(GU_TEXTURE_2D);
		sceGuClutMode(GU_PSM_8888, 0, 0xFF, 0); // 32-bit palette
		sceGuClutLoad((16 / 8), clut16); // upload 32*8 entries (256)
		sceGuTexMode(GU_PSM_T8, 0, 0, 0); // 8-bit image
		sceGuTexImage(0, 4, 4, 4, tex4_4);
		sceGuBlendFunc(GU_ADD, GU_SRC_ALPHA, GU_ONE_MINUS_SRC_ALPHA, 0, 0);
		sceGuTexFunc(GU_TFX_MODULATE, GU_TCC_RGB);
		sceGuTexFilter(GU_LINEAR, GU_LINEAR);
		sceGuTexScale(1.0f, 1.0f);
		sceGuTexOffset(0.0f, 0.0f);
		//sceGuAmbientColor(0xffffffff);
		//sceGuColor(0xffffffff);

		sceGuDrawArray(GU_SPRITES, GU_TEXTURE_8BIT | GU_VERTEX_8BIT | GU_TRANSFORM_2D, vertexCount, 0, vertices);
	}
	sceGuFinish();
	sceGuSync(0, 0);
	sceGuSwapBuffers();
	
	dumpPixels(0, 0, 4, 4);
}

u16 bezier_indices[16] = { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15 };
VertexType3 __attribute__((aligned(16))) bezier_vertices[16] = {
	{ 0xFFFFFFFF, -0.500000,  0.500000, 0.000000 },
	{ 0xFFEEEEEE, -0.166667,  0.500000, 0.000000 },
	{ 0xFFDDDDDD,  0.166667,  0.500000, 0.000000 },
	{ 0xFFCCCCCC,  0.500000,  0.500000, 0.000000 },
	{ 0xFFBBBBBB, -0.500000,  0.166667, 0.000000 },
	{ 0xFFAAAAAA, -0.166667,  0.166667, 0.000000 },
	{ 0xFF999999,  0.166667,  0.166667, 0.000000 },
	{ 0xFF888888,  0.500000,  0.166667, 0.000000 },
	{ 0xFF777777, -0.500000, -0.166667, 0.000000 },
	{ 0xFF666666, -0.166667, -0.166667, 0.000000 },
	{ 0xFF555555,  0.166667, -0.166667, 0.000000 },
	{ 0xFF444444,  0.500000, -0.166667, 0.000000 },
	{ 0xFF333333, -0.500000, -0.500000, 0.000000 },
	{ 0xFF222222, -0.166667, -0.500000, 0.000000 },
	{ 0xFF111111,  0.166667, -0.500000, 0.000000 },
	{ 0xFF000000,  0.500000, -0.500000, 0.000000 },
};

/*

struct Vertex
{
   unsigned int color;
   float x, y, z;
};

struct Vertex __attribute__((aligned(16))) vertices[1*3] =
{
       {0xFF0000FF, 0.0f, -50.0f, 0.0f}, // Top, red
       {0xFF00FF00, 50.0f, 50.0f, 0.0f}, // Right, green
       {0xFFFF0000, -50.0f, 50.0f, 0.0f}, // Left, blue
};

void testBezier() {
	int n;
	int vtype = GU_INDEX_16BIT | GU_COLOR_8888 | GU_VERTEX_32BITF | GU_TRANSFORM_2D;
	//int vtype = GU_INDEX_16BIT | GU_COLOR_8888 | GU_VERTEX_32BITF | GU_TRANSFORM_3D;

	sceGuDepthRange(65535,0);
	sceGuStart(GU_DIRECT, list);
	sceGuClear(GU_COLOR_BUFFER_BIT | GU_DEPTH_BUFFER_BIT);
	{
		sceGuDisable(GU_TEXTURE_2D);
		sceGuColor(0xff0000ff);
		
		sceGumMatrixMode(GU_PROJECTION);
		sceGumLoadIdentity();
		sceGumOrtho(0, 480, 272, 0, -1, 1);

		sceGumMatrixMode(GU_VIEW);
		sceGumLoadIdentity();
 
		sceGumMatrixMode(GU_MODEL);
		sceGumLoadIdentity();

		if (!(vtype & GU_TRANSFORM_2D)) {
			ScePspFVector3 scale = {100.0f, 100.0f, 100.0f};
			ScePspFVector3 translate = {100.0f, 100.0f, 0.0f};

			sceGumTranslate(&translate);
			sceGumScale(&scale);
		} else {
			//float size = 100.0f;
			float translate = 100.0f;
			//float translate = 0.0f;
			for (n = 0; n < 16; n++) {
				bezier_vertices[n].x = bezier_vertices[n].x * size + translate;
				bezier_vertices[n].y = bezier_vertices[n].y * size + translate;
				bezier_vertices[n].z = 0.0f;
			}
		}

		sceGuPatchDivide(2, 2);
		sceGuDrawBezier(vtype, 4, 4, bezier_indices, bezier_vertices);
		//sceGuDrawArray(GU_LINE_STRIP, vtype, 4 * 4, bezier_indices, bezier_vertices);
	}
	sceGuFinish();
	sceGuSync(0, 0);
	sceGuSwapBuffers();
}
*/

#define BUF_WIDTH (512)
#define SCR_WIDTH (480)
#define SCR_HEIGHT (272)
#define PIXEL_SIZE (4) /* change this if you change to another screenmode */
#define FRAME_SIZE (BUF_WIDTH * SCR_HEIGHT * PIXEL_SIZE)
#define ZBUF_SIZE (BUF_WIDTH * SCR_HEIGHT * 2) /* zbuffer seems to be 16-bit? */

void init() {
	sceKernelDcacheWritebackAll();

	sceGuInit();
	sceGuStart(GU_DIRECT, list);

	sceGuDrawBuffer (GU_PSM_8888, (void*)0, BUF_WIDTH);
	sceGuDispBuffer (SCR_WIDTH, SCR_HEIGHT, (void*)FRAME_SIZE,BUF_WIDTH);
	sceGuDepthBuffer((void *)(FRAME_SIZE * 2), BUF_WIDTH);
	sceGuOffset     (2048 - (SCR_WIDTH / 2),2048 - (SCR_HEIGHT / 2));
	sceGuViewport   (2048, 2048, SCR_WIDTH, SCR_HEIGHT);
	sceGuDepthRange (0xc350, 0x2710);
	sceGuScissor    (0, 0, SCR_WIDTH, SCR_HEIGHT);
	sceGuEnable     (GU_SCISSOR_TEST);
	sceGuFrontFace  (GU_CW);
	sceGuClear      (GU_COLOR_BUFFER_BIT | GU_DEPTH_BUFFER_BIT);
	sceGuFinish     ();
	sceGuSync       (0, 0);

	sceDisplayWaitVblankStart();
	sceGuDisplay(GU_TRUE);
}

int main(int argc, char *argv[]) {
	init();
	testVertexAlignment();
	testColorAdd();
	//testBezier();
	return 0;
}