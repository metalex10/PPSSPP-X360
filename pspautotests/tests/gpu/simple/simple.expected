testVertexAlignment
Struct Size: 12
FF000000,FF000000,FF000000,FF000000,
FF000000,0000,0000,FF000000,
FF000000,0000,0000,FF000000,
FF000000,0000,0000,FF000000,
testColorAdd
0000,0000,0000,0000,
0000,0000,0000,0000,
0000,0000,0000,0000,
0000,0000,0000,0000,
